src/gs-details-page.c

```diff
@@ -687,62 +703,89 @@ gs_details_page_refresh_screenshots (GsDetailsPage *self)
 		break;
 	}
 
-	/* set screenshots */
+	/* reset screenshots */
 	gs_container_remove_all (GTK_CONTAINER (self->box_details_screenshot_main));
-	if (screenshots->len == 0) {
-		gs_container_remove_all (GTK_CONTAINER (self->box_details_screenshot_thumbnails));
-		return;
-	}
-
-	/* set the default image */
-	ss = g_ptr_array_index (screenshots, 0);
-	ssimg = gs_screenshot_image_new (self->session);
-	gtk_widget_set_can_focus (gtk_bin_get_child (GTK_BIN (ssimg)), FALSE);
-	gs_screenshot_image_set_screenshot (GS_SCREENSHOT_IMAGE (ssimg), ss);
-
-	/* use a slightly larger screenshot if it's the only screenshot */
-	if (screenshots->len == 1) {
-		gs_screenshot_image_set_size (GS_SCREENSHOT_IMAGE (ssimg),
-					      AS_IMAGE_LARGE_WIDTH,
-					      AS_IMAGE_LARGE_HEIGHT);
-	} else {
-		gs_screenshot_image_set_size (GS_SCREENSHOT_IMAGE (ssimg),
-					      AS_IMAGE_NORMAL_WIDTH,
-					      AS_IMAGE_NORMAL_HEIGHT);
-	}
-	gs_screenshot_image_load_async (GS_SCREENSHOT_IMAGE (ssimg), NULL);
-	gtk_box_pack_start (GTK_BOX (self->box_details_screenshot_main), ssimg, FALSE, FALSE, 0);
-	gtk_widget_set_visible (ssimg, TRUE);
-
-	/* set all the thumbnails */
 	gs_container_remove_all (GTK_CONTAINER (self->box_details_screenshot_thumbnails));
-	if (screenshots->len < 2)
+
+	if (screenshots->len == 0)
 		return;
 
 	list = gtk_list_box_new ();
 	gtk_style_context_add_class (gtk_widget_get_style_context (list), "image-list");
 	gtk_widget_show (list);
 	gtk_box_pack_start (GTK_BOX (self->box_details_screenshot_thumbnails), list, FALSE, FALSE, 0);
+
 	for (i = 0; i < screenshots->len; i++) {
 		ss = g_ptr_array_index (screenshots, i);
+
+		/* we need to load the main screenshot only once if we're online
+		 * but all times if we're offline (to check which are cached and
+		 * hide those who aren't) */
+		if (is_offline || main_screenshot == NULL) {
+			GtkWidget *ssmain = gs_screenshot_image_new (self->session);
+			gtk_widget_set_can_focus (gtk_bin_get_child (GTK_BIN (ssmain)), FALSE);
+			gs_screenshot_image_set_screenshot (GS_SCREENSHOT_IMAGE (ssmain), ss);
+			gs_screenshot_image_set_size (GS_SCREENSHOT_IMAGE (ssmain),
+						      AS_IMAGE_NORMAL_WIDTH,
+						      AS_IMAGE_NORMAL_HEIGHT);
+			gs_screenshot_image_load_async (GS_SCREENSHOT_IMAGE (ssmain), NULL);
+
+			/* when we're offline, the load will be immediate, so we
+			 * can check if it succeeded, and just skip it and its
+			 * thumbnails otherwise */
+			if (is_offline &&
+			    !gs_screenshot_image_is_showing (GS_SCREENSHOT_IMAGE (ssmain)))
+				continue;
```

Comment:
Is it okay that we continue here without incrementing num_screenshots_loaded?  I'm guessing not.

Reply:
The screenshot has not been loaded (even if we called it asynchronously here, we're offline, so it should load any cached screenshots immediately), so it's correct to continue without incrementing the value.